### PR TITLE
Feature/fill profile

### PR DESCRIPTION
**Pull request by MishaFomenko**
## Describe your changes

- added endpoints to handle and save to the DB user info from fillProfile page

## JIRA

[Link](https://olga-team.atlassian.net/jira/software/projects/TP/boards/3/backlog?selectedIssue=TP-81&atlOrigin=eyJpIjoiNDYwMWQ3OWVhMjFkNGE4NmJjYmI3YTM0OWE0OGM3OTgiLCJwIjoiaiJ9)

### **General**

- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [ ] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [x] Date and time formats are on the constants
- [x] Functions are public only if it's used outside the class
- [x] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/)

### Backend

- [ ] Swagger documentation updated
- [x] Database requests are optimized and not redundant
- [ ] Unit tests written
- [x] use ConfigService instead of process.env
- [x] use transactions if there is a call chain that mutates data in different tables
- [x] use @index decorator for frequently requested data

### CODE REVIEW

File: src/modules/user/services/fillProfile.service.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import { ConfigService } from '@nestjs/config';
+import { PaymentMethod } from '@stripe/stripe-js';
+import { EErrorMessage } from 'src/common/enums/error-message.enum';
+import { UserRepository } from 'src/modules/repository/services/user.repository';
+import { AddressDto } from 'src/modules/user/dto/fillProfileDtos/address.dto';
+import { GeneralInfoDto } from 'src/modules/user/dto/fillProfileDtos/generalInfo.dto';
+import { SizesDto } from 'src/modules/user/dto/fillProfileDtos/sizes.dto';
+import Stripe from 'stripe';
+
+import { RoleDto } from '../dto/fillProfileDtos/role.dto';
```
Comment (eugeneiv1) [inline src/modules/user/services/fillProfile.service.ts]: absolute path

File: src/modules/user/services/fillProfile.service.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    if (user.filled_profile_step < fillProfileStep) {
+      user.filled_profile_step = fillProfileStep;
+    }
+
+    user.cloth_size = sizesDto.clothSize;
+    user.shoes_size = sizesDto.shoesSize;
+    user.jeans_size = sizesDto.jeansSize;
+    await this.userRepository.save(user);
```
Comment (eugeneiv1) [inline src/modules/user/services/fillProfile.service.ts]: para

File: src/modules/user/services/aws.service.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { PutObjectCommand, S3Client, S3ClientConfig } from '@aws-sdk/client-s3';
+import { Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+
+@Injectable()
+export class S3Service {
+  private client: S3Client;
+  constructor(private readonly configService: ConfigService) {
```
Comment (eugeneiv1) [inline src/modules/user/services/aws.service.ts]: para

File: src/modules/user/services/aws.service.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    const awsConfig = {
+      accessKey: this.configService.get<string>('AWS_ACCESS_KEY'),
+      secretKey: this.configService.get<string>('AWS_SECRET_KEY'),
+      bucket: this.configService.get<string>('AWS_S3_BUCKET_NAME'),
+      region: this.configService.get<string>('AWS_S3_REGION'),
+    };
+    const params: S3ClientConfig = {
```
Comment (eugeneiv1) [inline src/modules/user/services/aws.service.ts]: para

File: src/modules/user/services/aws.service.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      region: awsConfig.region,
+      credentials: {
+        accessKeyId: awsConfig.accessKey,
+        secretAccessKey: awsConfig.secretKey,
+      },
+    };
+    this.client = new S3Client(params);
```
Comment (eugeneiv1) [inline src/modules/user/services/aws.service.ts]: para

File: src/modules/user/services/aws.service.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  public async uploadFile(
+    file: Express.Multer.File,
+    itemType: 'avatar',
+    name: string,
+  ): Promise<string> {
+    const config = this.configService.get('aws');
+    const filePath = this.buildPath(itemType, name, file.originalname);
+    await this.client.send(
```
Comment (eugeneiv1) [inline src/modules/user/services/aws.service.ts]: para

File: src/modules/user/services/aws.service.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        accessKeyId: awsConfig.accessKey,
+        secretAccessKey: awsConfig.secretKey,
+      },
+    };
+    this.client = new S3Client(params);
+  }
+
+  public async uploadFile(
+    file: Express.Multer.File,
+    itemType: 'avatar',
```
Comment (eugeneiv1) [inline src/modules/user/services/aws.service.ts]: why string?

File: src/modules/user/services/aws.service.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    const filePath = this.buildPath(itemType, name, file.originalname);
+    await this.client.send(
+      new PutObjectCommand({
+        Key: filePath,
+        ACL: 'public-read',
+        Bucket: config.bucket,
+        Body: file.buffer,
+      }),
+    );
+    return `https://black-circle-faq-team.s3.us-east-2.amazonaws.com/${filePath}`;
```
Comment (eugeneiv1) [inline src/modules/user/services/aws.service.ts]: para

File: src/modules/user/services/user.service.ts [hunk 9]
```diff
@@ -72,6 +78,78 @@ export class UserService {
...
+
+    const customer = await stripe.customers.create({
+      payment_method: cardDto.id,
+      email: cardDto.billing_details.email,
+      invoice_settings: {
+        default_payment_method: cardDto.id,
+      },
+    });
+
+    const user = await this.userRepository.findOneBy({ id });
```
Comment (eugeneiv1) [inline src/modules/user/services/user.service.ts:100]: mb move this check at start? so we dont need even to call stripe if user doesnt exist?

File: src/modules/user/services/user.service.ts [hunk 10]
```diff
@@ -72,6 +78,78 @@ export class UserService {
...
+
+    if (!user) {
+      throw new HttpException(
+        EErrorMessage.USER_NOT_EXIST,
+        HttpStatus.BAD_REQUEST,
+      );
+    }
```
Comment (eugeneiv1) [inline src/modules/user/services/user.service.ts:97]: mb move check of existence to start? so we dont need even to call stripe if user doesnt exist?

File: src/modules/user/services/user.service.ts [hunk 11]
```diff
@@ -72,6 +78,78 @@ export class UserService {
...
+
+    await this.userRepository.save(user);
+  }
+
+  public async saveGeneralInfo({
+    generalInfoDto,
+    id,
+  }: {
+    id: string;
+    generalInfoDto: GeneralInfoDto;
```
Comment (eugeneiv1) [inline src/modules/user/services/user.service.ts:124]: mb move to type?

File: src/modules/user/user.controller.ts [hunk 12]
```diff
@@ -99,4 +101,26 @@ export class UserController {
...
   }
+
+  @ApiBearerAuth()
+  @UseGuards(JwtAuthGuard)
+  @Post(ERouteName.SAVE_GENERAL_INFO)
+  async saveGeneralInfo(
+    @Param('id', ParseUUIDPipe) id: string,
+    @Body() generalInfoDto: GeneralInfoDto,
+  ): Promise<void> {
+    return await this.userService.saveGeneralInfo({ id, generalInfoDto });
```
Comment (eugeneiv1) [inline src/modules/user/user.controller.ts]: why pass object? mb remove object and then You can remove object in service input?

File: src/modules/user/user.controller.ts [hunk 13]
```diff
@@ -99,4 +101,26 @@ export class UserController {
...
+  @ApiBearerAuth()
+  @UseGuards(JwtAuthGuard)
+  @Post(ERouteName.SAVE_CARD_INFO)
+  async saveCardInfo(
+    @Param('id', ParseUUIDPipe) id: string,
+    @Param('step') step: number,
+    @Body()
+    { cardDto }: { cardDto: PaymentMethod },
+  ): Promise<void> {
+    return await this.userService.saveCardInfo({ id, step, cardDto });
```
Comment (eugeneiv1) [inline src/modules/user/user.controller.ts]: same

File: src/modules/user/services/user.service.ts (hunk 10 above)
Comment (MishaFomenko) [reply src/modules/user/services/user.service.ts:97]: fixed

File: docker-compose.yml [hunk 14]
```diff
@@ -2,8 +2,8 @@ version: '3'
 
 services:
   db:
-    image: mysql:latest
-    command: ["mysqld", "--mysql-native-password=ON"]
+    image: mysql:5.6
```
Comment (eugeneiv1) [inline docker-compose.yml]: ?

File: src/entities/order.entity.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { Column, Entity } from 'typeorm';
+
+import { ETableName } from '../common/enums/table-name.enum';
+import { BaseEntity } from '../entities/models/base.entity';
```
Comment (AlexandraVoloshyna) [inline src/entities/order.entity.ts:4]: imports

File: docker-compose.yml (hunk 14 above)
Comment (MishaFomenko) [reply docker-compose.yml]: fixed

**Review by eugeneiv1 — APPROVED** [review]

**Pull request merged**